gateways/authorizenet/request_builders.go

```diff
@@ -109,3 +114,86 @@ func authentication(merchantName string, transactionKey string) MerchantAuthenti
 		TransactionKey: transactionKey,
 	}
 }
+
+func addL2L3Data(authRequest *sleet.AuthorizationRequest, authNetAuthRequest *CreateTransactionRequest) *CreateTransactionRequest {
+	if authRequest.Level3Data != nil {
+		lineItemString := buildLineItemsString(authRequest)
+		if lineItemString != nil {
+			authNetAuthRequest.TransactionRequest.LineItem = json.RawMessage(*lineItemString)
+		}
+
+		authNetAuthRequest.TransactionRequest.Tax = &Tax{
+			Amount:  strconv.FormatInt(authRequest.Level3Data.TaxAmount.Amount, 10),
+		}
+
+		authNetAuthRequest.TransactionRequest.Duty = &Tax{
+			Amount:  strconv.FormatInt(authRequest.Level3Data.DutyAmount.Amount, 10),
+		}
+
+		authNetAuthRequest.TransactionRequest.Shipping = &Tax{
+			Amount:  strconv.FormatInt(authRequest.Level3Data.ShippingAmount.Amount, 10),
+		}
+
+		authNetAuthRequest.TransactionRequest.Customer = &Customer{
+			Id: authRequest.Level3Data.CustomerReference,
+		}
+	}
+
+	if authRequest.ShippingAddress != nil {
+		authNetAuthRequest.TransactionRequest.ShippingAddress = &BillingAddress{
+			FirstName: 	authRequest.CreditCard.FirstName,
+			LastName: 	authRequest.CreditCard.LastName,
+			Company: 	common.SafeStr(authRequest.ShippingAddress.Company),
+			Address: 	authRequest.ShippingAddress.StreetAddress1,
+			City:      	authRequest.ShippingAddress.Locality,
+			State:    	authRequest.ShippingAddress.RegionCode,
+			Zip:       	authRequest.ShippingAddress.PostalCode,
+			Country:   	authRequest.ShippingAddress.CountryCode,
+		}
+	}
+
+	return authNetAuthRequest
+}
+
+// Authorize net converts json to XML before processing the request. This leads to weird scenarios like repeating json
+// fields. LineItems is one of them so we will build it as a raw string
+func buildLineItemsString(authRequest *sleet.AuthorizationRequest) *string {
+	// No more than 30 lineitems
+	lineItemCount := len(authRequest.Level3Data.LineItems)
+	if lineItemCount > 30 {
+		lineItemCount = 30
+	}
```

Comment:
nit: not needed anymore

Reply:
we still need it so we know when to not put a comma

Reply:
You could just replace the last comma with the closing bracket, but I'm fine with either way